support Map as MappingSource

hduelme opened the pull request:
I added support for `Map` as `MappingSource`. 
If only `Map<String, ?>` as a valid source is present, unmapped target warnings/errors are suppressed. This fixes #167 
I also added an inspection if the Map is raw or does  not have a String as a key, like described in [MapToBeanRawMapMapper](https://github.com/mapstruct/mapstruct/blob/main/processor/src/test/java/org/mapstruct/ap/test/frommap/MapToBeanRawMapMapper.java) and in [MapToBeanTypeCheckMapper](https://github.com/mapstruct/mapstruct/blob/main/processor/src/test/java/org/mapstruct/ap/test/frommap/MapToBeanTypeCheckMapper.java). For this I added two fixes, add `<String, String>` if the `Map` is raw and replace the key with `String` if the key is not of type `String`.

thunderhook reviewed (commented):
Hey, sorry it took so long. I had a look and just made a few nitpicks.

filiphr commented:
Thanks for your work on this @hduelme and thanks for the review @thunderhook.

I also had a look. I think that the inspection should be on the warning level by default, in order to be aligned to what we are doing in the processor. Apart from that I think we need to be a but more lax, e.g. in the `MapToBeanRawMapper` and `MapToBeanTypeCheckMapper` if you add a target parameter matching the map then it would work. Also if you have a look at `MapToBeanNonStringMapAsMultiSourceMapper` you'll see that it is possible to use non string map as a source if the source matches the target mapping.

hduelme commented on testData/inspection/UnmappedTargetPropertiesWithFromMapMapping.java:
```diff
@@ -0,0 +1,25 @@
+import java.util.Map;
+
+@Mapper
+interface SinglMapper {
```
Fixed typo

hduelme commented on src/main/resources/org/mapstruct/intellij/messages/MapStructBundle.properties:
```diff
@@ -19,6 +19,11 @@ inspection.not.null.checkable.property.source.used.with.default.property.title=C
+inspection.wrong.map.mapping.map.type.raw=Raw map used for mapping Map to Bean
+inspection.wrong.map.mapping.map.type.raw.set.default=Replace {0} with {0}<String, String>
+inspection.wrong.map.mapping.map.key=Key must be of type String for mapping Map to Bean
+inspection.wrong.map.mapping.map.key.change.to.String=Change key type to String
```
changed to lowercase `string`

hduelme commented on src/main/resources/org/mapstruct/intellij/messages/MapStructBundle.properties:
```diff
@@ -27,6 +32,8 @@ intention.more.than.one.source.property=Only use one source property
 intention.more.than.one.default.source.property=Only use one default source property
 intention.not.null.checkable.property.source.used.with.default.property=Remove default properties
 intention.java.expression.remove.unnecessary.whitespace=Remove unnecessary whitespaces
+intention.wrong.map.mapping.map.type.raw=Add type to Map for mapping Map to Bean
```
deleted

hduelme commented on testData/inspection/UnmappedTargetPropertiesWithFromMapMapping.java:
```diff
@@ -0,0 +1,25 @@
+
+    Target map(Map<String, String> source);
+}
+@Mapper
```
test added

hduelme commented:
@filiphr, @thunderhook, thanks for the feedback.
I've fixed the typos, removed the unused intention, and used lowercase `string` in the properties file.
I also switched the default level to `WARNING` for `FromMapMappingMapTypeInspection`. Additionally, I disabled the inspection if the name of the map parameter matches a target parameter or if the map parameter is used in a source mapping. The inspection only runs if the map parameter is the only source parameter and is not used for mapping directly.

For the `UnmappedTargetPropertiesInspection` I changed the check so that the inspection is disabled if the key of one map parameter is of type string. One thing I found during the investigation is that maps, arrays, and iterables are currently not recognized as source parameters when used as method parameters. The correct behavior would be to use the method parameter name as a source.

filiphr commented:
Thanks @hduelme. Sorry that it took me a bit to review it and merge it.

>One thing I found during the investigation is that maps, arrays, and iterables are currently not recognized as source parameters when used as method parameters. The correct behavior would be to use the method parameter name as a source.

I am not sure that I understand what you meant with this. However, feel free to create an issue and a fix if you know what the bug is.

hduelme commented:
No problem. Thanks for having a look at this @filiphr 
I will open a bug and provide a example for the problem.
